File: 19th/코딩테스트 심화반/week7/24480번_알고리즘 수업 - 깊이 우선 탐색 2.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

vector<int> adj[100001];
int visited[100001]; // visited에 방문 순서를 함께 저장

int cnt = 0; // 현재 방문 횟수

void dfs(int n) {
    int adj_cnt = adj[n].size(); // 인접 정점 개수
    for (int i = 0; i < adj_cnt; i++) {
        int adj_n = adj[n][i]; // 인접 정점
        if (visited[adj_n] == 0) { // 아직 방문하지 않은 경우
            cnt++;
            visited[adj_n] = cnt;

            dfs(adj_n);
        }
    }
}

int main() {
    int N, M, R;
    cin >> N >> M >> R;

    int n1, n2;
    for (int i = 0; i < M; i++) {
        cin >> n1 >> n2;
        adj[n1].push_back(n2);
        adj[n2].push_back(n1);
    }

    for (int i = 1; i <= N; i++) {
        // 모든 인접 정점을 내림차순으로 정렬
        sort(adj[i].begin(), adj[i].end(), greater<>()); 
    }

    cnt++;
    visited[R] = cnt; // 시작 정점의 방문 순서는 1

    dfs(R);

    // 방문 순서를 정점 순서대로 출력
    for (int i = 1; i <= N; i++) { 
        cout << visited[i] << "\n";
    }
}

File: 19th/코딩테스트 심화반/week7/24480번_알고리즘 수업 - 깊이 우선 탐색 2.py
import sys
sys.setrecursionlimit(300000)

n, m, r = map(int, input().split())

# 인접 리스트
adj = [[] for _ in range(n + 1)]
for _ in range(m):
    u, v = map(int, input().split())
    adj[u].append(v)
    adj[v].append(u)

# 인접 리스트 내 각 노드를 내림차순 정렬 (문제 조건에 맞춰 DFS 순서 지정)
adj = [sorted(row, reverse=True) for row in adj]

idx, cnt = 1, 1
visited = [0] * (n + 1)
visited[r] = 1

def dfs(n):
    global idx, cnt
    global visited

    for i in adj[n]:  # 현재 노드와 연결된 노드 탐색
        if not visited[i]:  # 방문하지 않았다면
            cnt += 1  # 방문 순서 증가
            visited[i] = cnt  # 방문 순서 기록
            dfs(i)  # 재귀 호출

dfs(r)

# 1번 노드부터 n번 노드까지 방문 순서 출력
for i in range(n):
    print(visited[i+1])